### DISCUSSION TITLE
error: no match for 'operator!='

### DISCUSSION THREAD

**Discussion by Vinit-Pandit** [sourcemeta/jsontoolkit#652] [Q&A] [answered]
for this code 

```cpp
#include <sourcemeta/jsontoolkit/jsonl.h>
#include <cassert>
#include <sstream>
#include <iostream>

int main()
{
    std::istringstream stream{
        "{ \"foo\": 1 }\n{ \"bar\": 2 }\n{ \"baz\": 3 }"};

    for (const auto &document : sourcemeta::jsontoolkit::JSONL{stream})
    {
        assert(document.is_object());
        sourcemeta::jsontoolkit::prettify(document, std::cout);
        std::cout << '\n';
    }
}
```

I am getting this error 

```
error: no match for 'operator!=' (operand types are 'sourcemeta::jsontoolkit::JSONL::const_iterator' {aka 'sourcemeta::jsontoolkit::ConstJSONLIterator'} and 'sourcemeta::jsontoolkit::JSONL::const_iterator' {aka 'sourcemeta::jsontoolkit::ConstJSONLIterator'})
   12 |     for (const auto &document : sourcemeta::jsontoolkit::JSONL{stream})
      |                                          
```

GCC version : 11.4.0

**Comment by Vinit-Pandit** [comment]
But tests for this are running without throwing errors, i dont know why ?
 @jviotti ?

**Comment by jviotti** [comment]
@MeastroZI On what OS are you on? 

- Can you try compiling your example with another compile? I.e. LLVM or MSVC? Do you get an issue there too?
- With GCC 11.4.0 on whatever OS you are on, does the JSON Toolkit test suite compile and run well, or do you get failures?
- How are you compiling the example above, precisely? i.e. you can show the commands you are using?

**Comment by Vinit-Pandit** [comment, marked as answer]
Ah, it seems the issue isn't with the compiler itself, but rather with the default C++ standard, which is set to 17 on my system running Ubuntu 22

here is the edited CMakeList.txt which work fine
```
cmake_minimum_required(VERSION 3.22)
project(HelloWorld)
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)

add_executable(hello hello.cpp)
add_subdirectory(deps/jsontoolkit)


target_link_libraries(hello PUBLIC sourcemeta::jsontoolkit::json)
target_link_libraries(hello PUBLIC sourcemeta::jsontoolkit::jsonschema)
target_link_libraries(hello PUBLIC sourcemeta::jsontoolkit::jsonpointer)
target_link_libraries(hello PUBLIC sourcemeta::jsontoolkit::jsonl)
```

**Comment by jviotti** [comment]
Interesting! A more backwards compatible approach would be to use the spaceship operator instead of implementing specific ones like we do now. I believe that would make it work on C++17 too. See https://github.com/sourcemeta/jsontoolkit/issues/509

**Comment by Vinit-Pandit** [comment]
Hmm, and one more thing: the [contains](https://en.cppreference.com/w/cpp/container/map/contains) method for maps is not available for the standard version less than 20. So, I'm also encountering an error for that

**Comment by jviotti** [comment]
Yep. The documentation (https://jsontoolkit.sourcemeta.com/) clarifies that C++20 or later is expected. Probably a lot more would be needed to make it work on C++17 which is likely not worth it
